src/main/java/com/gym/gymsystem/controller/TrainingController.java

```diff
@@ -0,0 +1,151 @@
+import jakarta.validation.Valid;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import java.time.LocalDate;
+import java.util.List;
+
+@RestController
+@RequestMapping(value = "/api/training")
+public class TrainingController {
+    private final TrainingService trainingService;
+    private final TraineeService traineeService;
+    private final TrainerService trainerService;
+    private final AuthorizationService authorizationService;
+
+    public TrainingController(TrainingService trainingService, TraineeService traineeService, TrainerService trainerService, AuthorizationService authorizationService) {
+        this.trainingService = trainingService;
+        this.traineeService = traineeService;
+        this.trainerService = trainerService;
+        this.authorizationService = authorizationService;
+    }
+
+    @GetMapping("/trainers/not-assigned")
+    public ResponseEntity<List<TrainerInfo>> getNotAssignedActiveTrainers(
+            @RequestParam("findUsername") String findUsername) {
+        if (!authorizationService.isAdmin() && !authorizationService.isTrainer() && !authorizationService.isAuthenticatedUser(findUsername)) {
+            throw new CustomAccessDeniedException("You do not have permission to get this trainee's not assigned trainers.");
+        }
+        List<TrainerInfo> trainerInfos = trainingService.getTrainersNotAssignedToTrainee(findUsername);
+        return ResponseEntity.ok(trainerInfos);
+    }
+
+    @PutMapping("/trainers")
+    public ResponseEntity<List<TrainerInfo>> updateTraineeTrainers(
+            @RequestParam("findUsername") String findUsername,
+            @RequestBody UpdateTraineeTrainersRequest request) {
+
+        if (!authorizationService.isAdmin()  && !authorizationService.isAuthenticatedUser(findUsername)) {
+            throw new CustomAccessDeniedException("You do not have permission to update this trainee's trainers.");
+        }
+        List<Training> updatedTrainers = trainingService.updateTraineeTrainersByUsername(findUsername, request.getTrainersUsernames());
+        List<TrainerInfo> trainerInfos = updatedTrainers.stream()
+                .flatMap(training -> training.getTrainers().stream())
+                .map(trainer -> new TrainerInfo(
+                        trainer.getUser().getUsername(),
+                        trainer.getUser().getFirstName(),
+                        trainer.getUser().getLastName(),
+                        trainer.getSpecializations().stream()
+                                .findFirst()
+                                .map(TrainingType::getTrainingTypeName).get()
+                ))
+                .distinct()
+                .toList();
+
+        return ResponseEntity.ok(trainerInfos);
+    }
+
+    @GetMapping("/trainee")
+    public ResponseEntity<List<TrainingInfo>> getTraineeTrainings(
+            @RequestParam("traineeUsername") String traineeUsername,
```

Comment:
same here if this is a required field and you want to return all trainings for a trainee then use path parameter